src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1167,17 +1140,196 @@ private boolean isSubtypeUncheckedInternal(Type t, Type s, boolean capture, Warn
                 return isSubtypeUncheckedInternal(t.getUpperBound(), s, false, warn);
             } else if (!s.isRaw()) {
                 Type t2 = asSuper(t, s.tsym);
-                if (t2 != null && t2.isRaw()) {
-                    if (isReifiable(s)) {
-                        warn.silentWarn(LintCategory.UNCHECKED);
+                if (t2 != null) {
+                    if (t2.isRaw()) {
+                        if (isReifiable(s)) {
+                            warn.silentWarn(LintCategory.UNCHECKED);
+                        } else {
+                            warn.warn(LintCategory.UNCHECKED);
+                        }
+                        return true;
                     } else {
-                        warn.warn(LintCategory.UNCHECKED);
+                        if (!isSameType(t2, s) &&
+                                t2.isReferenceProjection() == s.isReferenceProjection() &&
+                                structuralComparator.visit(t2, s)) {
+                            warn.warn(LintCategory.UNCHECKED);
+                            return true;
+                        }
+                        return false;
                     }
-                    return true;
                 }
             }
             return false;
         }
+        // where
+        StructuralTypeComparator structuralComparator = new StructuralTypeComparator();
+        class StructuralTypeComparator extends TypeRelation {
+            private Set<TypePair> cache = new HashSet<>();
+
+            @Override
+            public Boolean visitType(Type t, Type s) {
+                switch (t.getTag()) {
```

Comment:
In general - it seems like this routine has been brought in from isSubtype, which I think is wrong. This should borrow the most from isSameType. I believe in the case of this visitor, it would have been ok to simply return true if the two types have same tags? E.g. doesn't seem like adding the `switch` really adds much.

Reply:
agreed